pkg/controller/openapi/product_reconciler.go

```diff
@@ -0,0 +1,360 @@
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      objName,
+			Namespace: p.openapiCR.Namespace,
+		},
+		Spec: capabilitiesv1beta1.ProductSpec{
+			Name:               name,
+			SystemName:         systemName,
+			Description:        description,
+			ProviderAccountRef: p.openapiCR.Spec.ProviderAccountRef,
+		},
+	}
+
+	// Deployment
+	product.Spec.Deployment = p.desiredDeployment()
+
+	// Methods
+	product.Spec.Methods = p.desiredMethods()
+
+	// Mapping rules
+	mappingRules, err := p.desiredMappingRules()
+	if err != nil {
+		return nil, err
+	}
+	product.Spec.MappingRules = mappingRules
+
+	// backend usages
+	// current implementation assumes same system name for backend and product
+	backendSystemName := p.desiredSystemName()
+	product.Spec.BackendUsages = map[string]capabilitiesv1beta1.BackendUsageSpec{
+		backendSystemName: capabilitiesv1beta1.BackendUsageSpec{
+			Path: "/",
+		},
+	}
+
+	product.SetDefaults(p.Logger())
+
+	// internal validation
+	validationErrors := product.Validate()
+	if len(validationErrors) > 0 {
+		return nil, errors.New(validationErrors.ToAggregate().Error())
+	}
+
+	err = p.SetOwnerReference(p.openapiCR, product)
+	if err != nil {
+		return nil, err
+	}
+
+	return product, nil
+}
+
+func (p *ProductReconciler) productMutator(existingObj, desiredObj common.KubernetesObject) (bool, error) {
+	existing, ok := existingObj.(*capabilitiesv1beta1.Product)
+	if !ok {
+		return false, fmt.Errorf("%T is not a *capabilitiesv1beta1.Product", existingObj)
+	}
+	desired, ok := desiredObj.(*capabilitiesv1beta1.Product)
+	if !ok {
+		return false, fmt.Errorf("%T is not a *capabilitiesv1beta1.Product", desiredObj)
+	}
+
+	// Metadata labels and annotations
+	updated := helper.EnsureObjectMeta(existing, desired)
+
+	// OwnerRefenrence
+	updatedTmp, err := p.EnsureOwnerReference(p.openapiCR, existing)
+	if err != nil {
+		return false, err
+	}
+	updated = updated || updatedTmp
+
+	// Maybe too rough compare method?
+	// What if product controller adds or modifies something?
+	// the openapi controller will be reconciliating.
+	// maybe compare only "managed" fields
+	if !reflect.DeepEqual(existing.Spec, desired.Spec) {
+		diff := cmp.Diff(existing.Spec, desired.Spec)
+		p.Logger().Info(fmt.Sprintf("%s spec has changed: %s", common.ObjectInfo(desired), diff))
+		existing.Spec = desired.Spec
+		updated = true
+	}
+
+	return updated, nil
+}
+
+func (p *ProductReconciler) desiredSystemName() string {
+	// Same as backend system name
+	// Duplicated implementation. Refactor
+	if p.openapiCR.Spec.ProductSystemName != "" {
+		return p.openapiCR.Spec.ProductSystemName
+	}
+
+	return helper.SystemNameFromOpenAPITitle(p.openapiObj)
+}
+
+func (p *ProductReconciler) desiredObjName() string {
+	// DNS1123 Label compliant name. Due to UIDs are 36 characters of length this
+	// means that the maximum prefix lenght that can be provided is of 26
+	// characters. If the generated name is not DNS1123 compliant an error is
```

Comment:
Is this comment obsolete? the `desiredObjName` does not return an error

Reply:
the function does not return error, however it could return some name not compliant with DNS1123. 

The comments is there as a reminder that extra processing could be needed.